aiozk/recipes/shared_lock.py

```diff
@@ -1,27 +1,48 @@
-from aiozk import exc, Deadline
-
 from .base_lock import BaseLock
 
 
 class SharedLock(BaseLock):
-    async def acquire_read(self, timeout=None):
-        deadline = Deadline(timeout)
-        result = None
-        while not result:
+    class LockProxy:
+        def __init__(self, shared_lock, znode_label, wait_with_party):
+            self.shared_lock = shared_lock
+            self.znode_label = znode_label
+            self.wait_with_party = wait_with_party
+            self.is_locked = False
+            self.in_use = False
+
+        async def __aenter__(self):
+            await self.acquire()
+
+        async def __aexit__(self, exc_type, exception, tb):
+            await self.release()
+
+        async def acquire(self, timeout=None):
+            if self.in_use:
+                raise RuntimeError(f'called simultaneously')
+            self.in_use = True
             try:
-                result = await self.wait_in_line(
-                    "read", deadline.timeout, blocked_by=("write")
-                )
-            except exc.SessionLost:
-                continue
-        return result
-
-    async def acquire_write(self, timeout=None):
-        deadline = Deadline(timeout)
-        result = None
-        while not result:
+                await self.shared_lock.wait_in_line(
+                    self.znode_label,
+                    timeout=timeout,
+                    wait_with_party=self.wait_with_party)
+            except Exception as e:
+                self.in_use = False
+                raise e from None
+
+            self.is_locked = True
+
+        async def release(self):
             try:
-                result = await self.wait_in_line("write", deadline.timeout)
-            except exc.SessionLost:
-                continue
-        return result
+                await self.shared_lock.get_out_of_line(self.znode_label)
+            finally:
+                self.is_locked = False
+                self.in_use = False
+
+        def locked(self):
+            return self.is_locked
+
+    def __init__(self, base_path):
+        super().__init__(base_path)
+
+        self.reader_lock = self.LockProxy(self, 'rlock', True)
```

Comment:
Probably it would be better if we use names from recipe https://zookeeper.apache.org/doc/r3.1.2/recipes.html#Shared+Locks, eg. `read` and `write`. Other libraries, that follows the same naming would be compatible in that case. I don't really expect it, but it won't create more confusion.
Also, could you please put the link to it into comment

Reply:
I will change the names.